Type mismatch: Property IntValue.Value should be Int32, but it was Int64

A bit of background: I've been using [Lune](https://github.com/filiptibell/lune) to convert over some old files into a more usable file system format rather than arbitrary binary files. It seems like some of my older files aren't able to be deserialized.

I can reopen them and save again, and it seems to work fine after that, but I really don't want to do that another 2,000 times. If this is a non-issue, I totally understand lol. It's a very old file from early 2018.

Lune log with `RUST_LOG=trace`:

```lua
local inputFile = fs.readFile("repo.rbxl")
local game = roblox.deserializePlace(inputFile)
print(game.ClassName)
```

```txt
[TRACE rbx_binary::chunk] Chunk "PROP" (compressed: 24, len: 22, reserved: 0)
[TRACE rbx_binary::deserializer::state] PROP chunk (IntValue.Value, instance type 29, prop type 29
[TRACE rbx_binary::deserializer::state] Known prop, canonical name Value and type Int64, with None migration
Failed to read document from buffer - Type mismatch: Property IntValue.Value should be Int32, but it was Int64
```

Remodel log with `RUST_LOG=trace`:
```lua
local game = remodel.readPlaceFile("repo.rbxl")
print(game.ClassName)
```

```txt
[TRACE] Known prop, canonical name Tags and type Tags
[TRACE] Chunk "PROP" (compressed: 24, len: 22, reserved: 0)
[TRACE] PROP chunk (IntValue.Value, instance type 29, prop type 29
[TRACE] Known prop, canonical name Value and type Int64
[ERROR] callback error
        stack traceback:
                [C]: in field 'readPlaceFile'
                [string "remodel.lua"]:1: in main chunk
        caused by: Type mismatch: Property IntValue.Value should be Int32, but it was Int64
```

Repo file: https://github.com/ThatTimothy/rbx-binary-issue-repo/blob/main/repo.rbxl (didn't let me upload directly, sorry)

--- a/rbx_binary/src/deserializer/state.rs
+++ b/rbx_binary/src/deserializer/state.rs
@@ -489,6 +489,9 @@ impl<'a, R: Read> DeserializerState<'a, R> {
                         add_property(instance, &property, value.into());
                     }
                 }
+                // This branch allows values serialized as Int32 to be converted to Int64 when we expect a Int64
+                // Basically, we convert Int32 to Int64 when we expect a Int64 but read a Int32
+                // See: #301
                 VariantType::Int64 => {
                     let mut values = vec![0; type_info.referents.len()];
                     chunk.read_interleaved_i32_array(&mut values)?;

--- a/rbx_binary/src/deserializer/state.rs
+++ b/rbx_binary/src/deserializer/state.rs
@@ -489,6 +489,15 @@ impl<'a, R: Read> DeserializerState<'a, R> {
                         add_property(instance, &property, value.into());
                     }
                 }
+                VariantType::Int64 => {
+                    let mut values = vec![0; type_info.referents.len()];
+                    chunk.read_interleaved_i32_array(&mut values)?;
+
+                    for (value, referent) in values.into_iter().zip(&type_info.referents) {
+                        let instance = self.instances_by_ref.get_mut(referent).unwrap();
+                        add_property(instance, &property, (value as i64).into());
+                    }
+                }
                 invalid_type => {
                     return Err(InnerError::PropTypeMismatch {
                         type_name: type_info.type_name.clone(),

--- a/rbx_binary/src/deserializer/state.rs
+++ b/rbx_binary/src/deserializer/state.rs
@@ -538,6 +538,19 @@ impl<'a, R: Read> DeserializerState<'a, R> {
                         add_property(instance, &property, value.into());
                     }
                 }
+                // This branch allows values serialized as Float32 to be converted to Float64 when we expect a Float64
+                // Basically, we convert Float32 to Float64 when we expect a Float64 but read a Float32
+                // See: #301
+                VariantType::Float32 => {
+                    let mut values = vec![0.0; type_info.referents.len()];
+                    chunk.read_interleaved_f32_array(&mut values)?;
+
+                    for (value, referent) in values.into_iter().zip(&type_info.referents) {
+                        let instance = self.instances_by_ref.get_mut(referent).unwrap();
+                        let converted_value = f64::from(value);
+                        add_property(instance, &property, converted_value.into());
+                    }
+                }
                 invalid_type => {
                     return Err(InnerError::PropTypeMismatch {
                         type_name: type_info.type_name.clone(),

--- a/rbx_binary/src/deserializer/state.rs
+++ b/rbx_binary/src/deserializer/state.rs
@@ -498,7 +498,8 @@ impl<'a, R: Read> DeserializerState<'a, R> {
 
                     for (value, referent) in values.into_iter().zip(&type_info.referents) {
                         let instance = self.instances_by_ref.get_mut(referent).unwrap();
-                        add_property(instance, &property, (value as i64).into());
+                        let value_converted = i64::from(value);
+                        add_property(instance, &property, value_converted.into());
                     }
                 }
                 invalid_type => {

